### ISSUE TITLE
No repeat forever?

### ISSUE THREAD

**Issue by kentcb** [ReactiveX/rxdart#189]
Hi,

Possibly missing something obvious here, but I can't see a `repeat` without arguments. That is, repeat indefinitely. Does this exist?

**Comment by kentcb** [comment]
Woah, hang on. `repeat` in rxdart is not what I was expecting coming from Rx.NET. In Rx.NET, `Repeat()` will resubscribe to the source _when it completes_. `Repeat(int)` will do so only the specified number of times.

An hour of debugging later, I figured out that `repeat` in rxdart doesn't behave the same way - it just repeats all items it receives.

Worse, I can't seem to find an equivalent operator to Rx.NET's `Repeat` operator. Any ideas?

**Comment by frankpepermans** [comment]
Hmmm seems like ```Repeat``` is still one that we need to adapt, some ops like ```Repeat``` (or the ```ReplaySubject```) are a bit tricky with the Dart Streams to get right.

So we need a ```Repeat(int count, {bool indefinitely=false})```

**Comment by frankpepermans** [comment]
Quick branch to fix this: https://github.com/ReactiveX/rxdart/blob/transformers/repeat/lib/src/transformers/repeat.dart

I'll wrap it up once I get more time, so don't assume this is without issues :)

**Comment by kentcb** [comment]
@frankpepermans Thanks, but to be clear: this still has very different behavior to Rx.NET `Repeat`. The repeat there refers to re-subscribing to the source upon its completion, not recording values and repeating them a number of times after the source completes (as this rxdart operator does). It behaves like what the `retry` operator, except it "retries" regardless of whether it completes in error or not.

**Comment by frankpepermans** [comment]
No worries, again I'll finish this up later, was just a quick bit of work between other work

**Comment by frankpepermans** [comment]
@kentcb Resubscribing with Dart streams is problematic, because Dart Streams function differently from standard Rx Streams:
- In Dart, by default Streams are single subscription, which means that, if you resubscribe, you will get an RTE. The alternative is broadcast Streams, but they will not replay past events to new subscribers.

Now in the repeat operator, there is no way of controlling what type the initial StreamController really is, so the workaround here is to buffer and replay over the specified count.

Can you share a test where this solution breaks? I'll see what I can do with what we have

**Comment by kentcb** [comment]
@frankpepermans I'm at work right now, so can't provide my exact scenario. However, I was using `defer` to ensure the stream is being recreated:
```dart
Observable
  .defer(() => /* creates the inner observable */)
  .repeat(5);
```
What I want this to do:
1. Create the inner observable and subscribe to it
2. Wait until the observable completes, then re-subscribe. This should re-execute the `defer`, creating a new inner observable which is then subscribed

What it actually does:
1. Creates an inner observable and subscribes to it
2. Captures all the events from the inner observable
3. When the inner observable completes, replays those captured events 5 times

Are you saying that even with `defer`, a second subscription will raise an exception? That makes me wonder what `defer` would even be useful for...

**Comment by frankpepermans** [comment]
No, ```defer``` simply takes a ```StreamFactory``` and returns a ```Stream```, so you could for example use defer to:
- create a ```ReplaySubject```
- subscribe to the original ```Stream```
- add all events to that ```ReplaySubject```
- return the ```Stream``` of that ```ReplaySubject```

At this point, the deferred ```Stream``` can be subscribed to as much as you'd like.

The problem comes with calling defer over and over again on the original ```Stream```.
If this is a single subscription one for example, it can be listened to just once.
So how will you make defer return a new ```Stream``` each time containing the data you need for the replay, without buffering the events somewhere?

I guess my question really was, do you really need it to resubscribe on each loop? Do you need the ```onListen``` handler to fire each time? (on the original ```Stream``` that is). If not, then would a buffered version, such as the one I was using, suffice?

**Comment by kentcb** [comment]
OK, the differences between rxdart and Rx.NET are really confusing me then. Here's my actual use case:
```dart
Stream<dynamic> _throttledGeneralSettingsSave(Stream<dynamic> actions, EpicStore<AppState> store) {
  final observableActions = Observable<dynamic>(actions);

  // Repeatedly trap the first action that implies a save is required, then debounce any subsequent actions
  // before triggering a save. Also, start each of these with a pending action so the application logic can know
  // that a save is imminent.
  final updates =
      observableActions.where((dynamic action) => action is UpdateReferenceIdTemplate || action is UpdateCompanyName || action is UpdateCompanyId);
  final saveCompletions = observableActions.where((dynamic action) => action is SaveGeneralSettingsSucceeded || action is SaveGeneralSettingsFailed);

  // TODO: can I use exhaustMap?? https://pub.dartlang.org/documentation/rxdart/latest/rx/Observable/exhaustMap.html

  final saveDemarcations = Observable<dynamic>.defer(() =>
        updates
          .first
          .asObservable()
          .switchMap<dynamic>((dynamic _) =>
            observableActions
              .debounce(Duration(milliseconds: 2000))
              .map<dynamic>((dynamic _) => const SaveGeneralSettings())
              .takeUntil<dynamic>(saveCompletions.first.asObservable())
              .startWith(const SaveGeneralSettingsPending())));
      //.repeat(2^32);

  // Do the work of actually saving, producing a success or failure as a result.
  final saveRequests = observableActions.where((dynamic action) => action is SaveGeneralSettings);
  final saves = saveRequests.switchMap<dynamic>((dynamic _) => Observable<dynamic>.fromFuture(api
      .saveGeneralSettings(store.state.settingsState.generalSettings.dto)
      .then<dynamic>((dynamic _) => const SaveGeneralSettingsSucceeded())
      .catchError((Exception exception) => SaveGeneralSettingsFailed(ActionException(exception, const SaveGeneralSettings())))));

  // Tie the two observables together for our final observable.
  final result = Observable.merge<dynamic>([saveDemarcations, saves]);

  return result;
}
```
This is a redux epic. The comments explain things pretty well, but the idea is to detect when saves need to occur and debounce them. In addition, as soon as it's known that a save needs to occur, a `SaveGeneralSettingsPending` action is added so that the application can respond accordingly (e.g. show a "save pending..." message).

If I leave the `repeat` commented out, it works as expected, but only once. I want to repeat the inner observable so that future saves also trigger as expected. How would I go about achieving this with rxdart?

**Comment by kentcb** [comment]
I found a way to fake the behavior I want by abusing `retry`. If I change my `saveDemarcations` to this:
```dart
final saveDemarcations =
    Observable<dynamic>.retry(() => updates.first.asObservable().switchMap<dynamic>(
          (dynamic _) => Observable<dynamic>.concat([
                updates
                    .startWith(const SaveGeneralSettingsPending())
                    .debounce(Duration(milliseconds: 2000))
                    .map<dynamic>((dynamic _) => const SaveGeneralSettings())
                    .takeUntil<dynamic>(saveCompletions.first.asObservable()),
                Observable<dynamic>.error("fake failure"),
              ]),
        ));
```
then everything works. But obviously I have to fake an "error" just to get `retry` to behave how I would expect `repeat` to behave.

**Comment by MiYogurt** [comment]
user avatar use family photo is popular in this year? nice family. good luck for you-all。

**Comment by frankpepermans** [comment]
fixed in 0.20.0
